common/quotas/global/collection/internal/counted.go

```diff
@@ -74,34 +74,35 @@ func (c CountedLimiter) Wait(ctx context.Context) error {
 }
 
 func (c CountedLimiter) Reserve() clock.Reservation {
-	c.usage.idle.Store(0) // not idle regardless of the result
-
+	// caution: keep fallback limiter in sync!
+	// TODO: ideally it would actually be the same code, but that's a bit awkward due to needing different interfaces.
 	res := c.wrapped.Reserve()
 	if res.Allow() {
-		// may be used or canceled, return a wrapped version so it's tracked
-		// when we know which it is.
-		return countedReservation{
+		return allowedReservation{
 			wrapped: res,
 			usage:   c.usage,
 		}
 	}
-	// cannot be used, just count the rejection immediately
-	// and return the original so it's a bit cheaper
-	c.usage.rejected.Add(1)
+	// rejections cannot be rolled back, so they are always counted immediately
+	c.usage.Count(false)
+	res.Used(false)
 	return res
 }
 
 func (c CountedLimiter) Collect() UsageMetrics {
 	return c.usage.Collect()
 }
 
-func (c countedReservation) Allow() bool {
-	return c.wrapped.Allow()
+func (c allowedReservation) Allow() bool {
+	return true
 }
 
-func (c countedReservation) Used(wasUsed bool) {
+func (c allowedReservation) Used(wasUsed bool) {
+	c.wrapped.Used(wasUsed)
 	if wasUsed {
-		c.usage.allowed.Add(1)
+		// only counts as allowed if used, else it is hopefully rolled back.
+		// this may or may not restore the token, but it does imply "this limiter did not limit the event".
+		c.usage.Count(true)
 	}
```

Comment:
I originally started just always using the reservation-wrapper, but the semantics were kinda weird-feeling and it felt more error-prone the longer I looked at it:
```go
c.wrapped.Used(wasUsed)
if c.Allow() {
  if wasUsed {
    c.usage.allowed.Add(1)
  } // else not used
} else {
  c.usage.rejected.Add(1)
}
```
it's fine and it'd work, it just felt... odd.